### PR TITLE
Fix the bug that rolling-update throws error when using generateName

### LINKED ISSUE #13350: rolling-update: error: resource name may not be empty
I'm getting a `resource name may not be empty` error when performing a rolling update. From the looks of it, the update succeeded, but the old rc is never removed.

```
$ kubectl rolling-update consul-ef20k -f rc.json
```

```
Creating 
At beginning of loop: consul-ef20k replicas: 2,  replicas: 1
Updating consul-ef20k replicas: 2,  replicas: 1
At end of loop: consul-ef20k replicas: 2,  replicas: 1
At beginning of loop: consul-ef20k replicas: 1,  replicas: 2
Updating consul-ef20k replicas: 1,  replicas: 2
At end of loop: consul-ef20k replicas: 1,  replicas: 2
At beginning of loop: consul-ef20k replicas: 0,  replicas: 3
Updating consul-ef20k replicas: 0,  replicas: 3
At end of loop: consul-ef20k replicas: 0,  replicas: 3
error: resource name may not be empty
```

```
$ kubectl get pod
NAME                 READY     STATUS    RESTARTS   AGE
consul-kmekp-my90r   2/2       Running   0          4m
consul-kmekp-w7hw3   2/2       Running   0          5m
consul-kmekp-w9rht   2/2       Running   0          3m
```

```
$ kubectl get rc
CONTROLLER     CONTAINER(S)   IMAGE(S)                                   SELECTOR                    REPLICAS
consul-ef20k   consul         gliderlabs/consul-server:0.5               app=consul,version=latest   0
               kube2consul    eu.gcr.io/our-images/kube2consul:latest                               
consul-kmekp   consul         gliderlabs/consul-server:0.5               app=consul,version=v1.0.0   3
               kube2consul    eu.gcr.io/our-images/kube2consul:v1.0.0                               
```

```json
{
  "kind": "ReplicationController",
  "apiVersion": "v1",
  "metadata": {
    "generateName": "consul-",
    "labels": {
      "app": "consul"
    }
  },
  "spec": {
    "replicas": 3,
    "selector": {
      "app": "consul",
      "version": "latest"
    },
    "template": {
      "metadata": {
        "labels": {
          "app": "consul",
          "version": "latest"
        }
      },
      "spec": {
        "containers": [{
          "name": "consul",
          "image": "gliderlabs/consul-server:0.5",
          "imagePullPolicy": "Always",
          "command": ["/bin/consul", "agent", "-server", "-config-dir=/config", "-bootstrap-expect", "3", "-atlas", "company/production", "-atlas-join"],
          "lifecycle": {
            "preStop": {
              "exec": {
                "command": ["/bin/consul", "leave"]
              }
            }
          },
          "env": [{
            "name": "ATLAS_TOKEN",
            "value": "xxx"
          }]
        },
        {
          "name": "kube2consul",
          "image": "eu.gcr.io/our-images/kube2consul:latest"
        }]
      }
    }
  }
}
```

*(note that this json is dynamically updated - `latest` labels are changed to specific versions)*

### PR DESCRIPTION
**Pull request by janetkuo**
Fixes #13350.

Fixes the bug that `rolling-update` throws error when trying to get new rc when we don't specify name but use generateName.

### CODE REVIEW

**Comment by JanetKuo** [comment]
@k8s-bot test this please

**Comment by JanetKuo** [comment]
@k8s-bot test this please

**Comment by JanetKuo** [comment]
@k8s-bot test this please

File: pkg/kubectl/rolling_updater.go [hunk 1]
```diff
@@ -395,6 +397,12 @@ func (r *RollingUpdater) pollForReadyPods(interval, timeout time.Duration, oldRc
 // Existing controllers are validated to ensure their sourceIdAnnotation
 // matches sourceId; if there's a mismatch, an error is returned.
 func (r *RollingUpdater) getOrCreateTargetControllerWithClient(controller *api.ReplicationController, sourceId string) (*api.ReplicationController, bool, error) {
+	// Only create but not get when the controller has not name but generatedName,
+	// Otherwise it'll complain "no resource name".
+	if len(controller.Name) == 0 && len(controller.GenerateName) > 0 {
+		newRc, err := r.createTargetController(controller, sourceId)
```
Comment (thockin) [inline pkg/kubectl/rolling_updater.go]: This bypasses the check for Replicas <= 0 below - is that correct?

Could you instead do something like:

```
existing, err := testExists(controller)
if err != nil {
    //...
}
if controller.Spec.Replicas <= 0 {
    //...
}
newRc, err := r.createTargetController(controller, sourceId)
```

and

```
func testExists(controller) {
    if len(controller.Name) == 0 && len(controller.GenerateName) > 0 {
        return false, nil
    }
    return r.c.ReplicationControllers(controller.Namespace).Get(controller.Name)
}
```

Comment (ironcladlou) [inline pkg/kubectl/rolling_updater.go]: @thockin is correct- the replica count validation is required prior to creation.

Comment (JanetKuo) [inline pkg/kubectl/rolling_updater.go]: Fixed

**Comment by JanetKuo** [comment]
PTAL

**Comment by thockin** [comment]
LGTM

**Comment by JeanMertz** [comment]
@thockin we've updated to Kubernetes 1.1, and are still facing this exact same error

**Pull request merged**